src/main/java/io/a2a/server/tasks/TaskManager.java

```diff
@@ -0,0 +1,165 @@
+import static io.a2a.util.Assert.checkNotNullParam;
+
+import java.rmi.ServerError;
+import java.util.ArrayList;
+import java.util.List;
+
+import io.a2a.spec.Artifact;
+import io.a2a.spec.InvalidParamsError;
+import io.a2a.spec.Message;
+import io.a2a.spec.Part;
+import io.a2a.spec.Task;
+import io.a2a.spec.TaskArtifactUpdateEvent;
+import io.a2a.spec.TaskStatus;
+import io.a2a.spec.TaskStatusUpdateEvent;
+
+public class TaskManager {
+    private volatile String taskId;
+    private volatile String contextId;
+    private final TaskStore taskStore;
+    private final Message initialMessage;
+
+    public TaskManager(String taskId, String contextId, TaskStore taskStore, Message initialMessage) {
+        checkNotNullParam("taskStore", taskStore);
+        this.taskId = taskId;
+        this.contextId = contextId;
+        this.taskStore = taskStore;
+        this.initialMessage = initialMessage;
+    }
+
+    String getTaskId() {
+        return taskId;
+    }
+
+    String getContextId() {
+        return contextId;
+    }
+
+    public Task getTask() {
+        if (taskId == null) {
+            return null;
+        }
+        return taskStore.get(taskId);
+    }
+
+    public void saveTaskEvent(Task task) throws ServerError {
+        checkIdsAndUpdateIfNecessary(task.getId(), task.getContextId());
+        saveTask(task);
+    }
+
+    public void saveTaskEvent(TaskStatusUpdateEvent event) throws ServerError {
+        checkIdsAndUpdateIfNecessary(event.getTaskId(), event.getContextId());
+        Task task = ensureTask(event.getTaskId(), event.getContextId());
+
+
+        Task.Builder builder = new Task.Builder(task)
+                .status(event.getStatus());
+
+        if (task.getStatus().message() != null) {
+            List<Message> newHistory = task.getHistory() == null ? new ArrayList<>() : new ArrayList<>(task.getHistory());
+            newHistory.add(task.getStatus().message());
+            builder.history(newHistory);
+        }
+
+        task = builder.build();
+        saveTask(task);
+    }
+
+    public void saveTaskEvent(TaskArtifactUpdateEvent event) throws ServerError {
+        checkIdsAndUpdateIfNecessary(event.getTaskId(), event.getContextId());
+        Task task = ensureTask(event.getTaskId(), event.getContextId());
+
+        // Append artifacts
+        List<Artifact> artifacts = task.getArtifacts() == null ? new ArrayList<>() : new ArrayList<>(task.getArtifacts());
+
+        Artifact newArtifact = event.getArtifact();
+        String artifactId = newArtifact.artifactId();
+        boolean appendParts = event.getAppend() != null && event.getAppend();
+
+        Artifact existingArtifact = null;
+        int existingArtifactIndex = -1;
+
+        for (int i = 0; i < artifacts.size(); i++) {
+            Artifact curr = artifacts.get(i);
+            if (curr.artifactId() != null && curr.artifactId().equals(artifactId)) {
+                existingArtifact = curr;
+                existingArtifactIndex = i;
```

Comment:
Do we need a break here?

Reply:
Ah yes! This is one of the scenarios we need to test better by the way https://github.com/fjuma/a2a-java-sdk/issues/46